Bug fix - SetState callback called before component state is updated in ReactShallowRenderer

accordeiro opened the pull request:
This PR aims to fix the bug described in #11496

A test was created to reproduce the error, and to fix it, the `ReactShallowRenderer` callback calling logic was changed, so that it would stop calling the callbacks after the `enqueue*` functions, and would start calling them after finishing mounting or updating.

Please let me know if there any questions or if anything should be changed / improved :)

accordeiro commented:
Hmm something is going wrong with the tests, I'm gonna check this.

gaearon commented on packages/react-test-renderer/src/ReactShallowRenderer.js:
```diff
@@ -192,31 +194,41 @@ class ReactShallowRenderer {
+
+  _invokeCallback() {
+    if (typeof this._callback === 'function' && this._publicInstance) {
+      this._callback.call(this._publicInstance);
```
You'll probably want to immediately set it to null before calling. Since now it is outdated.

gaearon commented on packages/react-test-renderer/src/ReactShallowRenderer.js:
```diff
@@ -192,31 +194,41 @@ class ReactShallowRenderer {
+    this._publicInstance = null;
+  }
+
+  _updateCallback(callback, publicInstance) {
```
Maybe _enqueueCallback

accordeiro commented:
Thanks for the review @gaearon – sending a push with the changes in a bit. 

BTW, I'm not sure what's going on with the CircleCI tests, all of them seem to be passing by looking at the log. Is this a CI issue, or something I should address?

accordeiro commented:
I've just pushed the requested changes.

The CI tests are stating that I didn't run `prettier` for `packages/react-test-renderer/src/ReactShallowRenderer.js` (which I did). I ran a full `yarn prettier-all` and there are no suggested changes for `ReactShallowRenderer.js` either – what should I do?

gaearon commented:
Maybe you didn't run `yarn` first? The Prettier version has changed.

accordeiro commented:
Cool, thanks :) 

Looks like all checks have passed now.

gaearon commented on packages/react-test-renderer/src/ReactShallowRenderer.js:
```diff
@@ -183,6 +184,7 @@ class ReactShallowRenderer {
 
     if (shouldUpdate) {
```
Does ReactDOM call the callback if `shouldComponentUpdate` skips an update? I would expect so, but this PR doesn't. Can you verify this?

Maybe it's better to move the callback call [here](https://github.com/accordeiro/react/blob/11b931d3b5e386e2fe164d8c7b2494081beb661e/packages/react-test-renderer/src/ReactShallowRenderer.js#L105). Then you don't need to duplicate it in two branches.

accordeiro commented:
OK, I've moved the callback call.

Regarding the behavior when `shouldComponentUpdate` skips an update, I'm not sure how I can test this. I've written this following test:

```javascript
it('setState callback should be called even if update is skipped', () => {
    let stateSuccessfullyUpdated = false;

    class Component extends React.Component {
      constructor(props, context) {
        super(props, context);
        this.state = {
          hasUpdatedState: false,
        };
      }

      shouldComponentUpdate() {
        return false;
      }

      componentWillMount() {
        this.setState(
          {hasUpdatedState: true},
          () => stateSuccessfullyUpdated = this.state.hasUpdatedState,
        );
      }

      render() {
        return <div>{this.state.hasUpdatedState}</div>;
      }
    }

    const shallowRenderer = createRenderer();
    shallowRenderer.render(<Component />);
    expect(stateSuccessfullyUpdated).toBe(true);
  });
```
... and the callback is indeed being called, but I suspect this isn't enough for the code to even trigger a `shouldComponentUpdate()` check. I've thought about rendering the component, then triggering an event to start an update cycle, which would then call `setState`- but I'm afraid I'm overcomplicating things. Any suggestions? Thanks!

gaearon commented:
>OK, I've moved the callback call.

Note I don't *know* if that's how it works. Can you please check whether ReactDOM calls the callback in this case or not first?

gaearon commented:
>I've thought about rendering the component, then triggering an event to start an update cycle, which would then call setState- but I'm afraid I'm overcomplicating things.

You could call `getMountedInstance()` and then just call `setState` on it from outside.

accordeiro commented:
Just wanted to follow up and say I'm working on this today :)

accordeiro commented:
> Note I don't _know_ if that's how it works. Can you please check whether ReactDOM calls the callback in this case or not first?

It seems like ReactDOM does call the `setState` callback even if an update is skipped – I've just written some tests for both `ReactShallowRenderer` and `ReactDOM` to ensure their behavior is consistent. Do these tests make sense to you?

gaearon commented on packages/react-dom/src/__tests__/ReactDOM-test.js:
```diff
@@ -338,4 +338,45 @@ describe('ReactDOM', () => {
+
+      componentWillMount() {
+        setState = (newState, callback) => this.setState(newState, callback);
+        getState = () => this.state;
```
These helpers make it a bit hard to read what's really going on in the test. Could you please just store the instance instead, and then read `state` and call `setState` on it directly?

gaearon commented on packages/react-dom/src/__tests__/ReactDOM-test.js:
```diff
@@ -338,4 +338,45 @@ describe('ReactDOM', () => {
+    let setState, getState;
+
+    const div = document.createElement('div');
+    document.body.appendChild(div);
```
I don't think we need to add it to the body for this test.

gaearon commented on packages/react-test-renderer/src/ReactShallowRenderer.js:
```diff
@@ -192,31 +193,44 @@ class ReactShallowRenderer {
+    }
+  }
+
+  _invokeCallback() {
```
Maybe `_invokeCallbackIfNecessary`? It doesn't always exist.

accordeiro commented:
Done! Just pushed the requested changes.

gaearon commented on packages/react-dom/src/__tests__/ReactDOM-test.js:
```diff
@@ -338,4 +338,42 @@ describe('ReactDOM', () => {
+    expect(mockFn).not.toBeCalled();
+
+    instance.setState({hasUpdatedState: true}, () => {
+      expect(mockFn).toBeCalled();
```
This doesn't verify that the callback itself has been called.

accordeiro commented:
Awesome, thanks! :)

accordeiro commented:
OK, I'll investigate this. I will probably only be able to follow up later this week – is this ok?

gaearon commented:
Sure

accordeiro commented:
> This doesn't seem to handle multiple callbacks correctly.

You're right, I was simply replacing any previously set callback with the new one. I'm not sure if multiple callbacks is something that can happen in practice, but I've rewritten the code to make sure they all get called (and that the queue is cleaned up after they are called): https://github.com/facebook/react/pull/11507/commits/f242683caf88219dcc8c1595dfd4a8d53e091a0e

Please let me know if you think I should write a test for the multiple-callback case as well.

>  Did it start to do it in 16? Because if we don't do it in 16 either, we should probably change the shallow renderer to ignore them. Otherwise we should change the ReactDOM test to verify this specific case.

I've had a hard time finding out what actually happen on React 15. I've found [this issue](https://github.com/facebook/react/issues/1740), but it seems like the conclusion was that the callback wasn't guaranteed to be called.

In any case, it does seem like React 16 calls the `setState` callback inside `componentWillMount`:
 I've written a test to ensure behavior is consistent between `ReactShallowRenderer` and `ReactDOM`: https://github.com/facebook/react/pull/11507/commits/194ff74577c33ae5ec7ca28f0fdd562442adc546

gaearon commented on packages/react-test-renderer/src/ReactShallowRenderer.js:
```diff
@@ -193,24 +193,23 @@ class ReactShallowRenderer {
-      _callback.call(_publicInstance);
-    }
+  _invokeCallbacks() {
+    this._callbacks = this._callbacks.filter(({callback, publicInstance}) => {
```
Let's clear them early before calling them. So that if one throws, old callbacks are still removed from the list.

gaearon commented:
>Please let me know if you think I should write a test for the multiple-callback case as well.

Please do! Should be as easy as calling `setState` multiple times.

gaearon commented:
>I've had a hard time finding out what actually happen on React 15.

You can install React 15 and try to see what it does. :-)

gaearon commented:
I think you're right that todo was outdated.

accordeiro commented:
Thanks for the feedback! I'll work on this in the next couple of days and follow up.

accordeiro commented:
> Please do! Should be as easy as calling `setState` multiple times.

OK, I've added a test for the multiple callback case: https://github.com/facebook/react/pull/11507/commits/b3b078874597d572a53a9c9c87f0dd0b38b8b195

I'm not entirely happy with calling `setState` inside `componentWillMount` for this test, but it seems to cover the logic we want to test. What do you think?

> Let's clear them early before calling them. So that if one throws, old callbacks are still removed from the list.

I've updated the logic to clear the `_callbacks` attribute before actually invoking the callbacks: https://github.com/facebook/react/pull/11507/commits/2141b8cf8243954be16adcb54c90df256bb7ea2d

> I think you're right that todo was outdated.

Cool :) Do you still need me to install React 15 and test the behavior, though?

gaearon commented on packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js:
```diff
@@ -845,6 +845,100 @@ describe('ReactShallowRenderer', () => {
+        this.setState({foo: 'bar'}, () => mockFn());
+        this.setState({foo: 'foobar'}, () => mockFn());
+
+        callbackQueueLength = shallowRenderer._updater._callbacks.length;
```
We shouldn't be reading an internal field. Let's change this to only test for observable public behavior.

gaearon commented on packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js:
```diff
@@ -845,6 +845,100 @@ describe('ReactShallowRenderer', () => {
+    expect(callbackQueueLength).toBe(2);
+    expect(mockFn.mock.calls.length).toBe(2);
+
+    // Ensure the callback queue is cleared after the callbacks are invoked
```
Instead we should re-render, and ensure callbacks didn't get called again.

accordeiro replied:
OK, I've updated the test to ensure callbacks aren't called more than once.

accordeiro commented on packages/react-test-renderer/src/__tests__/ReactShallowRenderer-test.js:
```diff
@@ -890,8 +888,6 @@ describe('ReactShallowRenderer', () => {
         this.setState({foo: 'bar'}, () => mockFn());
         this.setState({foo: 'foobar'}, () => mockFn());
-
-        callbackQueueLength = shallowRenderer._updater._callbacks.length;
```
I didn't find a way to replace this specific test (that the callback queue was actually growing). However, the test that ensures that the `mockFn` was called twice – do you think this is enough?

gaearon replied:
Yes

accordeiro commented:
Thanks! I've re-written the commented parts.

gaearon commented:
Thanks for sticking through with it!

accordeiro commented:
Thanks! I had a great time and learned a lot :)
